File: db4oj.tests/src/com/db4o/db4ounit/optional/monitoring/QueryMonitoringSupportTestCase.java
/* Copyright (C) 2009  Versant Inc.   http://www.db4o.com */
package com.db4o.db4ounit.optional.monitoring;

import java.util.*;

import javax.management.*;

import com.db4o.config.*;
import com.db4o.diagnostic.*;
import com.db4o.foundation.*;
import com.db4o.monitoring.*;
import com.db4o.query.Query;

import db4ounit.*;
import db4ounit.extensions.OptOutNotSupportedJavaxManagement;
import db4ounit.extensions.fixtures.*;

@decaf.Remove(unlessCompatible=decaf.Platform.JDK15)
public class QueryMonitoringSupportTestCase extends QueryMonitoringTestCaseBase implements CustomClientServerConfiguration, OptOutNotSupportedJavaxManagement {
	
	@Override
	protected void configure(Configuration config) throws Exception {
		super.configure(config);
		config.add(new QueryMonitoringSupport());
	}

	public void configureClient(Configuration config) throws Exception {
	}

	public void configureServer(Configuration config) throws Exception {
		configure(config);
	}
	
	@Override
	protected Class<?> beanInterface() {
		return QueriesMBean.class;
	}
	
	@Override
	protected void store() throws Exception {
		db().store(new Item("foo"));
	}
	
	public void testClassIndexScan() throws Exception {
		exercisePerSecondCounter("ClassIndexScansPerSecond", new Runnable() { public void run() {
			triggerClassIndexScan();
		}});
	}
	
	public void testAverageQueryExecutionTime() {
		
		Assert.areEqual(0.0, bean().getAttribute("AverageQueryExecutionTime"));
		
		triggerQueryExecutionTime(1000);
		Assert.areEqual(1000.0, bean().getAttribute("AverageQueryExecutionTime"));
		
		triggerQueryExecutionTime(200);
		triggerQueryExecutionTime(500);
		Assert.areEqual(350.0, bean().getAttribute("AverageQueryExecutionTime"));
		
	}

	private void triggerQueryExecutionTime(final int executionTime) {
		final Query query = newQuery(Item.class);
		fileSession().callbacks().queryOnStarted(trans(), query);
		_clock.advance(executionTime);
		fileSession().callbacks().queryOnFinished(trans(), query);
	}
	
	public void testQueriesPerSecond() {
		
		final ByRef<Integer> queryMode = ByRef.newInstance(0);
		
		exercisePerSecondCounter("QueriesPerSecond", new Runnable() { public void run() {
			switch (queryMode.value.intValue() % 3) {
			case 0:
				triggerOptimizedQuery();
				break;
			case 1:
				triggerUnoptimizedQuery();
				break;
			case 2:
				triggerSodaQuery();
				break;
			}
			queryMode.value = queryMode.value.intValue() + 1;
		}});
	}
	
	public void testClassIndexScanNotifications() throws Exception {
		
		final List<Notification> notifications = startCapturingNotifications(classIndexScanNotificationType());
		
		triggerClassIndexScan();
		
		Assert.areEqual(1, notifications.size());
		
		final Notification notification = notifications.get(0);
		Assert.areEqual(classIndexScanNotificationType(), notification.getType());
		Assert.areEqual(Item.class.getName(), notification.getUserData());
	}

	private String classIndexScanNotificationType() {
		return LoadedFromClassIndex.class.getName();
	}

	private void triggerClassIndexScan() {
		final Query query = newQuery(Item.class);
		query.descend("_id").constrain("foo");
		query.execute().toArray();
	}
	
	private void triggerSodaQuery() {
		newQuery(Item.class).execute().toArray();
	}

	@Override
	protected String beanID() {
		return Db4oMBeans.mBeanIDForContainer(fileSession());
	}


}

File: db4o.net/Db4objects.Db4o.Tests/native/Db4objects.Db4o.Tests/CLI1/Monitoring/QueryMonitoringSupportTestCase.cs
/* Copyright (C) 2007   Versant Inc.   http://www.db4o.com */
#if !CF && !SILVERLIGHT
using System.Diagnostics;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Monitoring;
using Db4objects.Db4o.Monitoring.Internal;
using Db4objects.Db4o.Query;
using Db4oUnit;
using Db4oUnit.Extensions.Fixtures;

namespace Db4objects.Db4o.Tests.CLI1.Monitoring
{
	class QueryMonitoringSupportTestCase : QueryMonitoringSupportTestCaseBase, ICustomClientServerConfiguration
	{
		protected override void Configure(IConfiguration config)
		{
			config.Add(new QueryMonitoringSupport());
		}

		public void ConfigureServer(IConfiguration config)
		{
			Configure(config);
		}

		public void ConfigureClient(IConfiguration config)
		{
		}

        protected override void Db4oSetupAfterStore()
        {
            Container().ProduceClassMetadata(ReflectClass(typeof(Item)));
        }


		public void TestQueriesPerSecond()
		{
            using (PerformanceCounter counter = PerformanceCounterSpec.QueriesPerSec.PerformanceCounter(FileSession()))
            {
                Assert.AreEqual(0, counter.RawValue);

		        ExecuteGetAllQuery();
		        ExecuteClassOnlyQuery();
		        ExecuteOptimizedNQ();
		        ExecuteUnoptimizedNQ();

		        ExecuteOptimizedLinq();
		        ExecuteUnoptimizedLinq();
		        Assert.AreEqual(6, counter.RawValue);
            }
		}

		public void TestClassIndexScansPerSecond()
		{
			AssertCounter(
                PerformanceCounterSpec.ClassIndexScansPerSec.PerformanceCounter(FileSession()),
				ExecuteSodaClassIndexScan);
		}
		
		private void ExecuteClassOnlyQuery()
		{
			NewQuery(typeof(Item)).Execute();
		}

		private void ExecuteGetAllQuery()
		{
			NewQuery().Execute();
		}

		private void ExecuteSodaClassIndexScan()
		{
			IQuery query = NewQuery(typeof(Item));
			query.Descend("_id").Constrain(42);
			query.Execute();
		}
	}
}
#endif
